ui/loading_indicators.py

```diff
@@ -1,50 +1,163 @@
+
+Vocabulary:
+*Idle* word is used for classes that don't know how much processing is left to do
+Example : throbber, dotted loading screen
+*Progress* word is used for classes that know how much work is done and how much work is left.
+Progress classes have a property `progress` that is used by the app programmer to indicate the advancement
+of the task. (range [0-1])
+
+"""
+
+
+class CenteredTextRenderer(object):
+    def draw_centered_text(self, draw, content, device_size):
+        # type: (ImageDraw, str, tuple) -> None
+        w, h = draw.textsize(content)
+        dw, dh = device_size
+        coords = (dw / 2 - w / 2, dh / 2 - h / 2)
+        draw.text(coords, content, fill=True)
+
 
 class LoadingIndicator(Refresher):
 
     def __init__(self, i, o, *args, **kwargs):
+        Refresher.__init__(self, self.on_refresh, i, o)
+        self._progress = 0
         Refresher.__init__(self, self.on_refresh, i, o, *args, **kwargs)
         self.t = None
 
     def on_refresh(self):
-        raise NotImplementedError
+        pass
 
     def run_in_background(self):
         if self.t is not None or self.in_foreground:
             raise Exception("LoadingIndicator already running!")
         self.t = Thread(target=self.activate)
-        self.t.daemon=True
+        self.t.daemon = True
         self.t.start()
 
 
-class DottedProgressIndicator(LoadingIndicator):
+class ProgressIndicator(LoadingIndicator):
+
+    @property
+    def progress(self):
+        return float(self._progress)
+
+    @progress.setter
+    def progress(self, value):
+        self._progress = clamp(value, 0, 1)
+        self.refresh()  # doesn't work ? todo : check out why
+
+
+# =========================concrete classes=========================
+class IdleCircular(LoadingIndicator):
+
+    def __init__(self, i, o, *args, **kwargs):
+        LoadingIndicator.__init__(self, i, o, *args, **kwargs)
+        self.refresh_interval = 0.01
+        self._current_angle = 0
+        self._current_range = 0  # range or width of the throbber
+        self.rotation_speed = 360  # degree per second
+        self.counter = Counter()  # We use a counter to make the animation independent of the refresh-rate
+        self.start_time = 0
+
+    def activate(self):
+        self.start_time = time()
+        self.counter.start()
+        return Refresher.activate(self)
+
+    def refresh(self):
+        self.update_throbber_angle()
+        c = canvas(self.o.device)
+        c.__enter__()
+        x, y = c.device.size
+        radius = min(x, y) / 4
+        draw = c.draw
+        draw.arc(
+            (
+                x / 2 - radius, y / 2 - radius,
+                x / 2 + 1 + radius, y / 2 + radius
+            ),
+            start=(self._current_angle - self._current_range / 2) % 360,
+            end=(self._current_angle + self._current_range / 2) % 360,
+            fill=True
+        )
+        self.o.display_image(c.image)
```

Comment:
The canvas usage looks great! Will take this code into account when sketching out the ZPUI canvas functions, thank you!